### PR TITLE
Make MetadataFieldMapper extend ParametrizedFieldMapper

### PR DESCRIPTION
**Pull request by romseygeek**
This commit cuts over all metadata field mappers to parametrized format.  Several
of these mappers are non-configurable, and we add a new `FixedTypeParser`
that simplifies parsing and building of these mapper types.

### CODE REVIEW

File: plugins/mapper-size/src/main/java/org/elasticsearch/index/mapper/size/SizeFieldMapper.java [hunk 1]
```diff
@@ -19,88 +19,71 @@
...
 import java.util.Map;
 
 public class SizeFieldMapper extends MetadataFieldMapper {
     public static final String NAME = "_size";
 
     public static class Defaults  {
-        public static final EnabledAttributeMapper ENABLED_STATE = EnabledAttributeMapper.UNSET_DISABLED;
-
-        public static final FieldType SIZE_FIELD_TYPE = new FieldType();
+        public static final boolean ENABLED = false;
```
Comment (nik9000) [inline plugins/mapper-size/src/main/java/org/elasticsearch/index/mapper/size/SizeFieldMapper.java:38]: Do we need the `Defaults` object at all any more?

File: plugins/mapper-size/src/main/java/org/elasticsearch/index/mapper/size/SizeFieldMapper.java [hunk 2]
```diff
@@ -19,88 +19,71 @@
...
 
         @Override
         public MetadataFieldMapper getDefault(ParserContext context) {
-            return new SizeFieldMapper(Defaults.SIZE_FIELD_TYPE, Defaults.ENABLED_STATE,
+            return new SizeFieldMapper(Defaults.ENABLED,
                 new NumberFieldMapper.NumberFieldType(NAME, NumberFieldMapper.NumberType.INTEGER));
         }
     }
 
-    private EnabledAttributeMapper enabledState;
```
[the remark this answers is not in the document]
Comment (nik9000) [reply plugins/mapper-size/src/main/java/org/elasticsearch/index/mapper/size/SizeFieldMapper.java:98]: :+1:

File: plugins/mapper-size/src/main/java/org/elasticsearch/index/mapper/size/SizeFieldMapper.java [hunk 3]
```diff
@@ -129,37 +112,15 @@ public void parse(ParseContext context) {
 
     @Override
     protected void parseCreateField(ParseContext context) {
-        if (!enabledState.enabled) {
+        if (enabled == false) {
```
Comment (nik9000) [inline plugins/mapper-size/src/main/java/org/elasticsearch/index/mapper/size/SizeFieldMapper.java:115]: :+1:

File: server/src/main/java/org/elasticsearch/index/mapper/IdFieldMapper.java [hunk 4]
```diff
@@ -277,19 +263,13 @@ public void preParse(ParseContext context) throws IOException {
 
     @Override
     protected void parseCreateField(ParseContext context) throws IOException {
-        if (fieldType.indexOptions() != IndexOptions.NONE || fieldType.stored()) {
-            BytesRef id = Uid.encodeId(context.sourceToParse().id());
-            context.doc().add(new Field(NAME, id, fieldType));
-        }
+        BytesRef id = Uid.encodeId(context.sourceToParse().id());
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/index/mapper/IdFieldMapper.java:266]: I wonder if ever allowed disabling indexing the id! Oh well, :+1:

File: server/src/main/java/org/elasticsearch/index/mapper/MetadataFieldMapper.java [hunk 5]
```diff
@@ -45,25 +45,63 @@
         MetadataFieldMapper getDefault(ParserContext parserContext);
     }
 
-    @SuppressWarnings("rawtypes")
-    public abstract static class Builder<T extends Builder<T>> extends FieldMapper.Builder<T> {
-        public Builder(String name, FieldType fieldType) {
-            super(name, fieldType);
+    public static class FixedTypeParser implements TypeParser {
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/index/mapper/MetadataFieldMapper.java:48]: Can add javadoc?

**Review by nik9000 — COMMENTED** [review]
I left some small things, but I like it!

File: server/src/main/java/org/elasticsearch/index/mapper/CompletionFieldMapper.java [hunk 6]
```diff
@@ -138,7 +138,7 @@ private static CompletionFieldMapper toType(FieldMapper in) {
                 b.startArray(n);
                 c.toXContent(b, ToXContent.EMPTY_PARAMS);
                 b.endArray();
-            });
+            }, ContextMappings::toString);
```
[the remark this answers is not in the document]
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/index/mapper/CompletionFieldMapper.java:141]: It's a change in the signature of `setSerializer` - we need to serialize things to string as well, for conflict reporting.  I should add some specific tests to ParametrizedMapperTests for this as well though.

**Comment by romseygeek** [comment]
This ended up being a little more complicated, because of the way MetadataFieldMappers are merged: if a metadata mapper has nothing specified in a mapping update, then we keep the old mapper, which means that we need to distinguish between 'this parameter has a default value because it wasn't set' (don't serialize) and 'this mapper has a default value that was explicitly set and is changing the existing value' (do serialize).  The existing `Explicit<>` class is handy for this, but it needs to also have `equals/hashCode` implementations so that we can compare against defaults.  Ready for another review.

File: server/src/main/java/org/elasticsearch/index/mapper/CompletionFieldMapper.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/index/mapper/CompletionFieldMapper.java:141]: I've added a specific test here

File: server/src/main/java/org/elasticsearch/index/mapper/ParametrizedFieldMapper.java [hunk 7]
```diff
@@ -105,7 +108,7 @@ protected final void mergeOptions(FieldMapper other, List<String> conflicts) {
     }
 
     @Override
-    public final XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
```
[the remark this answers is not in the document]
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/index/mapper/ParametrizedFieldMapper.java]: ++, it's gone now

**Review by nik9000 — APPROVED** [review]
LGTM

**Pull request merged**